daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionTraversal.scala

```diff
@@ -29,11 +53,80 @@ object TransactionTraversal {
       initialToVisit = tx.getRootsList.asScala.view
         .map(RawTransaction.NodeId(_) -> Set.empty[Ref.Party])
         .to(FrontStack)
-      _ <- go(f, txVersion, nodes, initialToVisit)
-    } yield ()
+    } yield { (txVersion, nodes, initialToVisit) }
+  }
 
   @tailrec
-  private def go(
+  private def traverseWitnessesWithPackages(
+      txVersion: TransactionVersion,
+      nodes: Map[String, Node],
+      toVisit: FrontStack[(RawTransaction.NodeId, Set[Ref.Party])],
+      packagesWithParties: Map[String, Set[Ref.Party]] = Map.empty,
```

Comment:
Nit: as this is a mapping, I'd suggest hinting at that with a name like `packageToParties` or `packageToPartySet` or similar.